proposals/2024-04-10-native-support-for-info-metrics-metadata.md

```diff
@@ -0,0 +1,154 @@
+    In the latter case, however, an error has to be returned to the user.
+    The collision can be resolved by constraining the labels via data label matchers.
+    And of course, the user always has the option to go back to the original join syntax (or, even better, avoiding ingesting conflicting info metrics in the first place).
+* Simplify enriching of query results with info metric data (non-identifying) labels in PromQL, e.g. via a new function, based on aforementioned TSDB API.
+* Ensure backwards compatibility with current Prometheus usage.
+* Minimize potential conflicts with existing metric labels.
+
+### Audience
+
+Prometheus maintainers.
+
+## How
+
+* Track the info metric's identifying label set over time (in case it changes) - storage model details to be elaborated in separate proposal.
+* Keep info metric indexes in storage - storage model details to be elaborated in separate proposal.
+  * Info metric indexes maintaining per info metric the different identifying label sets it has had over its lifetime.
+  * Indexing the different identifying label sets an info metric has had over its lifetime allows determining which are potential matches for a given metric, before considering the time dimension.
+* Add a method to the TSDB API for matching info metric data labels to a time series, given a certain timestamp and potentially data label matchers - the method will use the aforementioned info metric indexes.
+  * Candidate info metrics are found by searching info metric indexes for info metrics with identifying labels contained in the input label set.
+    * Each candidate info metric's identifying label set is obtained _for the timestamp in question_ - storage model details to be elaborated in separate proposal.
+    * If that identifying label set is not a match, the info metric is ignored.
+    * If several info metrics with the same name are found, the one with the latest sample is chosen (i.e., older metrics are considered stale).
+  * Data labels are picked from the found info metrics according to the rules defined in the Goals section.
+    * Each info metric's data labels are determined by taking those of the metric's labels which are not in the identifying label set.
+* Augment the OTLP endpoint to specify `target_info`'s identifying labels when ingesting write requests.
+* Simplify the inclusion of info metric labels in PromQL through a new `info` function: `info(v instant-vector[, ls label-selector])`.
+  This function will be UI for the aforementioned TSDB API.
+
+Using the `info` function, we can simplify the previously given PromQL join example as follows:
+
+```
+sum by (k8s_cluster_name, http_status_code) (
+  info(
+    rate(http_server_request_duration_seconds_count[2m]),
+    {k8s_cluster_name=~".+"}
+  )
+)
+```
+
+TODO:
+
+* Specify detection of info metric identifying labels for other ingestion methods than OTLP.
+* Define how this functionality would work together with OOO samples.
+
+## Alternatives
```

Comment:
This might have been discussed before, but I wonder if we are missing the simplest solution of query rewriting as a considered alternative. Given the example query on line [119](https://github.com/prometheus/proposals/pull/37/files#diff-e74174dfba7ed37042ee0c8fef56c9fad81cab4fc448067c3477488b1136fba0R119), can we rewrite it into a join and let the engine execute it with the existing semantics of PromQL. Information needed from the join can either be defined implicitly or specified by the user as arguments.

Reply:
The `info` function can't be simply reduced to syntactic sugar for joins though, for the following reasons at least:

1. The `info` function automatically finds suitable info metrics to join with (the MVP only supports `target_info`, but we aim to lift that limitation in the future)
1. The `info` function automatically knows which are info metrics' identifying labels, i.e. which to join on (the MVP assumes `instance` and `job`, but we aim to lift that limitation in the future)
1. The `info` function solves the problem of temporary conflicts between info metrics due to non-identifying labels changing, until the old version goes stale

> Information needed for the join can either be implicitly deducted or specified by the user as arguments.

Assuming we were to use query rewriting, how would you deduce which info metric to join against (when `info` is no longer limited to just the `target_info` metric)? If the user, as in join queries, has to supply which info metric to join with and which labels to join on plus which labels to include, have we really gained much versus status quo?

Reply:
I can only see this working well for a static list of well known info metrics like.. `target_info` or by specifying which info metrics to use expicitely. If i have multiple info metrics that match by accident and dont specify matchers I probably get confusing labels added. Thats not possible if i specify the info metric explicitely or constrain it to a well known list like `target_info`. Prometheus is too flexible to make such convenience safe I think.

Reply:
> I can only see this working well for a static list of well known info metrics like.. target_info or by specifying which info metrics to use expicitely.

@MichaHoffmann if that is your concern with the `info` function (and @fpetkovski's?), I think that should be an explicit criticism of the proposed design, rather than suggesting `info` as proposed here can be simply reduced to syntactic sugar. If you think the design should be explicit about picking info metrics rather than automatic (although that can be filtered down via the `__name__` label matcher), I think there should be an alternative proposal. I'm not sure though if it's most feasible to list competing proposals in the same document, or whether they should be separate.

Reply:
Oh, Sorry I didnt want to propose something differently, I was just concerning out loud. What I want to say is that the name label matcher for info metric probably should be requirement for this to be generally least surprising for the user.

Reply:
Thanks, I can note your wish about the name label matcher being a requirement, so a consensus can be reached.

Reply:
I should note that my original idea was indeed that there is generally a relatively low number of info metrics so that the default matching behavior is "all the info metrics known to the TSDB", which is in fact one of the reasons that makes the "joining" easier.

Note that with a "general metadata store", you would probably also end up with "give me all the metadata you know for this metric", which is (more or less) equivalent to "give me all the info metrics that have an identifying labels match with this metric".